### PR TITLE
Fix calldatacopy wrapround

### LINKED ISSUE #289: Incorrect Return Value Compared to Geth
Another one discovered via differential fuzzing:

`cabal run test -- --quickcheck-replay=783978 -p "random-contract-concrete-call". reproduced at `dbe984a96042a3873ecdffb84c60151d6229689a`

### PR DESCRIPTION
**Pull request by msooseth**
## Description
This fixes/improves:
* wraparound bug in `copySlice`
* adds test cases for `CallDataCopy` to `Tracing.hs`
* improves fuzzing capability of `Tracing.hs`
* improves printing of `ByteString`-s via helper function `byteStringToHex`
* minor refactors `Tracing.hs` so `checkTraceAndOutputs` is a separate function, usable in test cases

## Checklist

- [x] tested locally
- [x] added automated tests
- [ ] updated the docs
- [x] updated the changelog

Closes #289

### CODE REVIEW

**Comment by msooseth** [comment]
Nice, so this breaks some yul tests due to:
```
<interactive>: Internal error: TODO: implement copySlice with a symbolically sized region
```

Likely issue is the change in copy from:
```
      copySlice xOffset yOffset size bs mem
```

To:
```
      copySlice xOffset yOffset (Expr.min size (bufLength bs)) bs mem
```

Which is more correct I believe, but likely `bufLegth bs` is symbolic

**Review by d-xo — APPROVED** [review]
lgtm, just one small comment

File: src/EVM/Types.hs [hunk 1]
```diff
@@ -1308,6 +1308,16 @@ formatString bs =
...
 
+-- To display ByteStrings as HEX values
+byteStringToHex :: ByteString -> String
+byteStringToHex bs = foldMap (++"") $ "0x" : map toHex (BS.unpack bs)
+
+toHex :: Word8 -> String
+toHex w = case showHex w "" of
+           [w1,w2] -> [w1, w2]
+           [w2]    -> ['0', w2]
+           _       -> "showHex returned []"
```
[the remark this answers is not in the document]
Comment (msooseth) [reply src/EVM/Types.hs]: Oh, good point. I replaced it with simpleHex :) Thanks!

**Review by d-xo — APPROVED** [review]
looks good!

**Pull request merged**